RFC: 79 CDK v2.0 (#156)
aws/aws-cdk-rfcs: text/0079-cdk-2.0.md

---
feature name: cdk-2.0
start date: 2020-3-29
rfc pr: https://github.com/aws/aws-cdk-rfcs/pull/156
related issue: https://github.com/aws/aws-cdk-rfcs/issues/6
---

<!-- replace the blockquoted sections with your content -->
# Summary

CDK v2.0!

This RFC details the strategy for building and releasing aws-cdk 2.0, in addition to changes to tooling required to support this strategy. For details on the specific features themselves, see the corresponding RFCs.

# Motivation

Since the CDK was announcned as "generally available", the team has tried to limit breaking changes for users. However, a handful of features have emerged that the core team believes will make the CDK significantly simpler to use. These changes require a 2.0 release to signal to users that code changes may be required for adoption.

> The main change of note is to consolidate all of the aws service construct libraries into a single package. The motivation for this change is detailed in [the corresponding RFC](https://github.com/aws/aws-cdk-rfcs/issues/6).

Reviewer: ```suggestion
The main change of note is to consolidate all of the aws service construct libraries into a single package. The motivation for this change is detailed in [the corresponding RFC](https://github.com/aws/aws-cdk-rfcs/blob/master/text/0006-monolothic-packaging.md).
```

> The main change of note is to consolidate all of the aws service construct libraries into a single package. The motivation for this change is detailed in [the corresponding RFC](https://github.com/aws/aws-cdk-rfcs/issues/6).

Reviewer: ```suggestion
The main change of note is to release the CDK as a single package instead of multiple package. The motivation for this change is detailed in [the corresponding RFC](https://github.com/aws/aws-cdk-rfcs/issues/6).
```

# Basic Example

### New Module Structure

*Installing aws-cdk 2.0*
```bash
npm install aws-cdk aws-cdk-lib
```

*Importing modules from aws-cdk 2.0*
```typescript
import {
  aws_ecs as ecs,
  aws_lambda as lambda,
  aws_s3 as s3,
} from 'aws-cdk-lib';
```

*Importing experimental constructs from aws-cdk 2.0*
```typescript
import {aws_appsync as appsync_experimental} from 'aws-cdk-lib/experimental'
```

# Design Summary

CDK 2.0 consists of the following notable changes:

- [Monolithic Packaging](https://github.com/aws/aws-cdk-rfcs/issues/6) - Combine all aws construct libraries into a single package.
- [Reset all Feature Flags](https://github.com/aws/aws-cdk-rfcs/issues/55) - Delete all defined feature flags and extraneous logic for the "old" state.
- Remove Deprecated APIs - Delete all constructs/methods marked with `@deprecated` flag.
- Separate Experimental Code - Extract experimental APIs into a separate entrypoint.

# Detailed Design

## Release Stages

> CDK v2.0 will be released in stages. The stages also dictate if and how the v1 modules continue to be published.

Reviewer: Maybe provide an overview of the stages before diving into them to give readers the full picture upfront

4. Generate module `package.json` - Create the package.json for the build of `monocdk-experiment`. This has no dependencies and includes `constructs` and `@types/node` as devDependencies.
5. Install dependencies - Run `npm install`.
6. Compile with jsii
7. Stage for local consumption - copies the compiled `.js` files from the build directory back to the repository in the `packages/monocdk-experiment/staging` directory. This allows local modules that depend on `monocdk-experiment` to build against it.
8. Package - module is packaged as a normal jsii package using `jsii-pacmak` in the repositories packaging step using `pack.sh`.

All `@aws-cdk/` namespaced libraries will continue to be published as normal.

### 2. Pre-release

`monocdk-experiment` is renamed to `aws-cdk-lib`. The version of this module now becomes v2.0 and is tagged as a "pre-release" version. This has varying formats for different language targets.

#### NodeJS
version number: `2.0.0-beta.1`
deploying: `npm publish --tag beta`
installing: `npm install aws-cdk-lib@beta`
package.json:
```sh

> "dependencies": {

Reviewer: Worth mentioning that this is for apps.  Library vendors should use `peerDependencies` here, correct?

Author: Yeah I'll add a section for library authors here.

  "aws-cdk-lib": "~2.0.0-beta.1"
}
```

> **Note: `~` ranges match all versions with the same major, minor, and patch numbers with a greater prerelease number. For example, the above would match `2.0.0-beta.2`. but not `2.0.1-beta.2`.**

Reviewer: Will it also match `2.1.1-beta.2`?

Reviewer: What's our strategy for bumping these numbers? I am wondering if perhaps the last number should reflect the major+minor of the corresponding 1.x?

Can we do something like this:

|v1|maps to v2|
|-|-|
|1.32.0|2.0.0-beta.32.0|
|1.33.2|2.0.0-beta.33.2|

Author: it wont match `2.1.1-beta.2`. It only matches pre-releases with the same major, minor, and patch with a newer prelease tag. So it has to be `2.0.0-beta.x`.

We will have to test if we can version this way, matching the numbers of the pre-release tag to the minor.patch numbers of the current v1 release. I agree that would be ideal but looking through npm semver docs I don't see examples of additional numbers beyond the `-beta.x`.

Reviewer 2: Spec says it's okay: https://semver.org/#spec-item-9

#### Java
version number: `2.0.0-beta-1`
pom.xml:
```xml
<dependencies>
    <dependency>
        <groupId>software.amazon.awscdk</groupId>
        <artifactId>lib</artifactId>
        <version>2.0.0-beta-1</version>
    </dependency>
</dependencies>
```

> 

Reviewer: again, what's the strategy for mapping v1 versions to these and how can users take a "caret" version?

Author: This is just meant to show how the version of the node module translates to the targets.

Maven does allow defining ranges including prerelease numbers (which I thought it didn't). Will add.

#### Dotnet
version number: `2.0.0-beta.1`
.csproj:
```xml
  <ItemGroup>
    <PackageReference Include="Amazon.CDK.Lib" Version="2.0.0-beta.1" />
  </ItemGroup>
```

#### Python
version numbeer: `2.0b1`
requirements.txt: 
```
aws-cdk.lib==2.0b1
```

All `@aws-cdk/` namespaced libraries will continue to be published with current `1.x.x` versions.

### Stable

> After a period of testing and gathering feedback from users, `aws-cdk-lib` will be promoted to a stable release with the version `2.0.0`. All other modules, excluding those under the `@aws-cdk/` namespace, will also get promoted to v2.0, though will remain backwards compatible with `@aws-cdk/` modules.

Reviewer 2: > All other modules, [...] will also get promoted to v2.0, though will remain backwards compatible with `@aws-cdk/` modules.

I would like to see a table with a list of modules (collapsing all the service packages into one), and what treatment they'll be getting. I'm afraid a couple of these things won't be like the others.

All `@aws-cdk/` namespaced libraries will continue to be published with current `1.x.x` versions.

### v1 Deprecation

`@aws-cdk/` namespaced modules will continue to receive non-breaking changes and be released regularly.

> When v1 module support is discontinued will ultimately be dictated by the introduction of some feature or change that the team agrees is worth the cost of dropping support and requires us to do so. Alternatively, if the burden of maintaining v1 module support is no longer warranted by the number of users continuing to rely on them.

Reviewer 3: I'm not convinced we're doing the right thing here.

> dictated by the introduction of some feature or change that the team agrees is worth the cost of dropping support and requires us to do so

I'd rather us plan out when this happens, rather than pivot this on some feature/change that may or may not happen in the future.

> if the burden of maintaining v1 module support is no longer warranted by the number of users continuing to rely on them

The way I'm reading this is - "not in any reasonable timeline". We've this happen with releases other in other tools or ecosystem (e.g., it took a long time python2.7 could be deprecated).

We should a take a different approach instead - define and announce a sunset period and/or a maintenance period for 1.0 and stop releases after that. 
We should also consider an upgrade tool for customers to move from 1.0 to 2.0 (similar to the monocdk rewrite?) in all our supported languages.

Another benefit we will get from this is that deprecated code and packages can actually be deleted from our code base, and we'll no longer have to maintain backwards compat.

Reviewer: I tend to agree and also becoming a bit concerned about our ability to actually hide deprecated APIs from v2.0 without really modifying the code. As part of #7708 and #7736 I ran into a few situations where stable APIs extend an experimental API (e.g. `assets.IAsset`). Apart from the fact that I think we should probably prohibit this, it will be impossible to hide `IAsset` even though it's marked `deprecated`. We have quite a few places like that in our code and I don't know how this will actually work.

An additional case is the constructs compatiblity layer, which will also require some actual code changes to get rid of.

I am leaning more and more towards a period of transitioning from a v1.0 code base to a v2.0 code base in which we can actually do some clean ups and leave the old stuff behind some how. This, naturally, will require us to plan better for this transition and be explicit about what we intend to backport. It also means that we may need to continue to release v1.x for a while from a branch.

Author: I'm leaning towards agreeing with both of you. I think we should agree on a period of continuing support for v1 modules and announce that up front. Additionally the cost of hiding deprecated APIs during build time appears to become more complex the more we look into it.

We could implement whatever we can with the v2 modules building from the v1 modules. Then we can branch the two, without creating a git branch, by copying all of the v1 modules src into the `aws-cdk-lib` modules src directory and change the way we are building to just build from those local copies. Then we can implement breaking changes in the `aws-cdk-lib` module and continue to release v1 as is. This lets us go back and add any critical updates to v1 modules, but obviously ends new features and minor bugfixes etc for v1 during the overlap period.

Reviewer: I am concerned that duplicating the sources inside our repo and restructuring of our source repo will introduce risk and quite a lot of work into the critical path of v2.0. It will also require an awfully manual process to back port anything from 2.0 to 1.x and vice verse. 

I would also consider a git branch as a mechanism to maintain these two version lines temporarily. Perhaps we can even do that at an early stage in order to break the version number: branch off to a 2.x branch _today_ and have it continuously merge from master and build monocdk like today. Then, update the version number there to 2.x and create a separate pipeline that will release it.

Then, we can start applying changes and continue to merge from master until we officially release 2.x and perform the switch.

Reviewer 4: Would something like this be feasible to keep it simple? Rather than trying to navigate a single branch driving two different release versions, we pick a date for the transition to a v2.0 release candidate, say the 'pre-release' date suggested above, and then on that date:

1. Create a v1.x branch of the repo. This will be the branch that the v1.x release(s) will be built from, and will only have security patches and important bugfixes backported into it. The v1.x releases of the CLI and Construct library are made off of the v1.x branch, and only the v1.x branch. This lets v1.x sticks around for a while for those that want to stay on it, but doesn't get any new features.

2. Mainline development shifts to a release-candidate process for v2.0. All new code targets the mainline branch. The CDK development team releases a new release candidate every N days/weeks, and when it looks stable enough for prime-time it does the official v2.0 release.

3. The reviewers for PRs on mainline would have to be cognizant of identifying changes that would need to be backported to the v1.x line, and there would have to be a process for that.

New v1.x releases are made as-needed, but their cadence should be drastically reduced as the vast majority of changes will be in mainline.  We don't break any existing apps that are out in the world because the v1.x line has a clean break from the v2.0 line.

#### [JSII Prerelease Versioning Awareness]()

In order to support releasing the `aws-cdk-lib` with a prerelease version number, JSII needs to parse the version number of the module its building and recognize it as a "prerelease" and convert it to whatever format is equivalent in the target languages package manager.

#### [JSII Submodule Support](https://github.com/aws/jsii/issues/1286)

JSII currently doesn't handle namespacing of exports. Support for this will be needed to allow building the `aws-cdk-lib` module in all of the cdk supported target languages. Though the initial work is complete, some open questions remain about the scope of this feature. Namely, in order to make migrating to the new module structure as easy as possible, [namespace renaming]() could potentially allow C#, Python, and Java users to not have to change the structure of their import statements.

#### [`aws-cdk-lib` Version Unlocking]()

Since `aws-cdk-lib` will have a version number differing from the one used by lerna to version all of the other packages in the repo until the "stable" phase, changes to the align-version and bump scripts will have to be made in the repo to handle this.

> #### [Deprecation Aware Builds]()

Reviewer: I wonder if this is worth a separate RFC or at least a deeper dive/prototype.

Author: I think a prototype is whats needed.

> A strategy for selectively including or excluding apis marked `deprecated` during build time. This can be performed at the JSII level and potentially can be linked to specific version numbers with additional information in the `deprecated` annotation. JSII can warn when APIs marked for deprecation within a specific major version release are being included in the build. For example:

Reviewer 3: What about deprecated packages?

Reviewer: Ideally we should omit them from the monocdk, but I think it might be impossible given we have some stable APIs that depend on deprecated/experimental APIs.

An api annotated like so:
```typescript
  /**

>   * @deprecated-v1

Reviewer 3: This feels more complicated than needed. 
Why not just maintain only 2 active versions at any time? Before we move to the next major version, all items marked removed from the n-1 version is removed from source code.

Author: Thats what the design was meant to present. We would still want a # next to each deprecation to know if it was deprecated during the current, or previous cycle right?

  */
  function myDeprecatedApi() {}
```

When inside a module versioned `1.x.x`
`jsii` - builds normally, no warning

when inside a module versioned `2.x.x` or greater
`jsii` - build fails with error stating that deprecated APIs are included and points to the corresponding lines of code. It will suggest usage of `--no-deprecations` option to exlcude these APIs.

> `jsii --no-deprecations` - builds without warning, but excludes the above method and others marked deprecated during the v1.0 cycle

Reviewer 3: Sorry, I'm not following everything that is said here. Can you clarify better on the text here? It might read better if you start with the use case and then state the flags to use for that use case.

Are you saying that there is now an option to continue to keep deprecated APIs in the new version?

Author: No the `--no-deprecations` option is what you use when you want to remove deprecated APIs from the build you are performing. When you build without it, and JSII sees deprecated APIs from the `current - 1` version cycle, it warns you saying that you are in fact including APIs that shouldn't be.

In addition, since the v1 modules will continue to be supported for some time, and the `aws-cdk-lib` package will be made of up these modules, the inclusion of any of these candidate features would need to be implemented in a way that didn't affect the v1 modules. How this would be done has not been determined, but may be affected by the nature of any particular change.

#### Recommended

The following are good candidates for inclusion in this major version release

- [Easier Identification of Experimental Modules and APIs](https://github.com/aws/aws-cdk-rfcs/issues/116)

There have been long running discussions about separating experimental APIs into a separate module to make them a more explicit "opt in" feature. Users would either have to install a different module (i.e. `aws-cdk-lib-exerimental`) or import from a different entrypoint (`import * as s3 from 'aws-cdk-lib/experimental`). Since this could be performed at build time and be an optional setting thats turned on or off, it could be introduced in the 2.0 release without affecting the v1 modules, this seems like an ideal time to implement. The technical details of how to achieve this are yet to be determined.

> 

Reviewer: what about stable APIs that extend/use experimental APIs?

Author: Yeah we have to decide if that counts as stable or experimental for these purposes. Experimental feels right but does that remove too much useful stuff for the user? We should run some experiments to get metrics for how frequent stable apis calls experimental ones.

Reviewer: By definition those APIs are stable (and jsii-diff will shout if you try to change them). I think we should add a validation to the jsii compiler that prohibits this (and switch all of those APIs to stable so we can enable this check from now on)

Reviewer 2: * Using: it's *super* useful to be able to add an experimental field to a stable struct. If we say that stable functions are not allowed to take structs with experimental members in them, we close this useful avenue. I don't think we should take the route of prohibiting that.

There is something to be said for the struct *itself* not being experimental, but given that the struct itself can be stable and literally every field in it could still be experimental, I'm not sure how useful it would be to have that additional restriction.

* Extending: I was all ready to say that an API class/interface not be able to extend an experimental class/interface, but after thinking about the previous issue I'm not so sure about that anymore. Inheriting from an experimental base just means that you'll inherit some experimental members. It should behave somewhat the same as if you just hand-declared a bunch of experimental members, and that shouldn't be disallowed, so why should this?

I guess there's something to be said about the stability of the class/interface name.

But on the other hand, if an interface of a class `ISomething` is experimental, then a consumer shouldn't be declaring variables of type `ISomething` to store that class instance on. Declare variables to be `const x: SomeConcreteSublass = gimmeObject();` instead.

All in all, I'm not so sure I have a problem with this anymore.

To my mind, feels like in the "stable" rendition of the library, jsii should behave as if the experimental types weren't `export`ed (same as it does now for private base classes and private base interfaces).

Reviewer 2: What I'm more concerned about is the naming of types.

Types should have the same FQNs whether they are experimental or stable. If we don't have that, switching a class from experimental to stable breaks all consumers, and that would be shit.

- [Public S3 Artifacts](https://github.com/aws/aws-cdk-rfcs/issues/39)

> A significant downside of packaging all of the aws service construct libraries together is that the size of dependencies shipped to the users in most cases will be increased significantly. Being able to remove assets, like custom resource lambda code, that are currently bundled with the modules, could mitigate some of the impact on package size. This could be accomplished by packaging said assets, uploading them to S3, and then creating publically accessible lambda function layers referencing them across all AWS regions. Whether v2 of the cdk could accomplish this without impacting v1 has not been determined.

Reviewer 2: It would be quite scary to me to have this maintenance burden of a set of public artifacts.

All of a sudden we'd be responsible for replicating these all across the world, have monitoring on them, have backups, have a release pipeline, ...

Reviewer 3: Can we get some numbers?

- [Rename ID Parameter](https://github.com/aws/aws-cdk/issues/3203)

> `id` is a reserved keyword in python. Renaming this parameter in constructs prevents possible errors. This likely could be implemented as a feature flag with the flag removed as part of the normal major version release process.

Reviewer 2: It's not a reserved keyword though, it's a built-in function. It might be a valid concern, but I find the linked ticket light on details and full of FUD.

> I think this could lead to potentials errors down the line for users that use the Python variant of CDK, since the id parameter could clash with the id() function.

Potential confusion how? What is the code he sees people typing that is wrong?

What's more likely is that his editor is yellow-underlining this `id` variable and saying "you shouldn't do this". 

That's fair enough, and might be a good enough reason to change it for that alone, but let's be accurate about why we're doing it: "prevent IDE linter warning on all constructs" is a lot sharper and easier to reason about than "something might go wrong somewhere".

> #### Not Recommended

Reviewer 2: Agreed on all counts. This is going to be hard enough already, let's not make it worse than necessary.

If a deprecated API isn't removed from a version that it was inteded to be, and users start relying on the presence of that API, then by fixing the bug and removing the API in future versions we potentionally introduce breaking changes that weren't intended by us or expected by the users.

Because of this reason, automated verification of our different builds and the presence/absence of APIs and feature flags is, as always, of high importance.

The cost of migration for users varies depending on the state of the application being migrated. All users upgrading to v2 will have to perform the following steps:

1. Update Dependencies - Changing from referencing the old `@aws-cdk/` packages to the new `aws-cdk-lib` package in the corresponding target language's package manifest (package.json, .sln, pom.xml, setup.py).

> 2. Update Imports - Since the current module structure is being changed, import statements in user's code will need to change to accommodate the new structure. This will be automated wherever needed with the [import rewriter](#import-rewriter).

Reviewer 2: Just a niggle, but an automatic rewriter is not guaranteed to work in the presence of removing deprecated APIs.

If you were using deprecated stuff, it's gone now. No way to rewrite around that.

Users may have to make additional code changes when upgrading if they are relying on deprecated APIs or feature flags that are being removed. This cost can be minimized through good documentation and information in the [migration guide](#migration-guide).

# Rationale and Alternatives

## Rationale

> The main motivation for moving forward with a 2.0 release now is the desire to start moving towards the single module structure. The current multi module structure is known to cause a lot of confusion for users. For more details on why aws-cdk is moving to a single package, see the [MonoCDK RFC](https://github.com/aws/aws-cdk-rfcs/issues/6).

Reviewer: It's not only user confusion or a "belief" that this is better. It's basically impossible to use CDK modules as peer dependencies since every transitive dependency change will be breaking.

Author: Will rethink the wording here. Mostly just trying to say "go read the mono rfc".

> This belief that releasing a single package will make user's lives easier is countered by a strong desire to limit the exposure of users to breaking changes. Though the aws-cdk is relatively young, one of the core tenants of the project is stability. Major version upgrades are an important part of building and maintaining libraries. However, limiting the cost of breaking changes included in major releases is important. It isn't in the interest of the project to drop a new major version, immediately stop supporting the previous version, and leave the onus on the user to "catch up".

Reviewer 2: ```suggestion
This belief that releasing a single package will make user's lives easier is countered by a strong desire to limit the exposure of users to breaking changes. Though the aws-cdk is relatively young, one of the core tenets of the project is stability. Major version upgrades are an important part of building and maintaining libraries. However, limiting the cost of breaking changes included in major releases is important. It isn't in the interest of the project to drop a new major version, immediately stop supporting the previous version, and leave the onus on the user to "catch up".
```

Regardless, the point of using this change to push towards a 2.0 release, is that the single package structure is "the new cdk". Meaning, the long term plan is to only release the `aws-cdk-lib` package and no longer publish the `@aws-cdk/` namespaced ones. The [MonoCDK RFC](https://github.com/aws/aws-cdk-rfcs/issues/6) has more info on the problems currently experienced by users caused by the current packaging strategy and these problems will always exist unless we stop supporting the `@aws-cdk/` packages.

Since we want to eventually supplant the current module structure with the new one, semver says we have to do it with the release of a major version.

# Adoption Strategy

All users will perform the following to adopt cdk v2

1. Update Dependencies - Changing from referencing the old `@aws-cdk/` packages to the new `aws-cdk-lib` package in the corresponding target language's package manifest (package.json, .sln, pom.xml, setup.py).

> 2. Update Imports - Since the current module structure is being changed, import statements in user's code will need to change to accommodate the change. This will be automated with the [import rewriter](#import-rewriter).

Reviewer: Hopefully not in all languages

Users may have to make additional code changes when upgrading if they are relying on deprecated APIs or feature flags that are being removed. A [migration guide](#migration-guide) will be written with details about each individual feature flag and deprecated API that users may need to handle differently when upgrading.

# Unresolved questions

The bulk of unresolved questions are listed in the [detailed design](#detailed-design) section. To recap:

1. What additional features beyond adding the necessary tooling for releasing a major version going forward, should be included if any?
2. As part of jsii submodule support, can namespace renaming mitigate the need for users of non-typescript target languages to restructure their import statements?
3. Can `experimental` APIs be separated at build time to make their usage more explicitly "opt-in" for users?

> 4. With regards to automatic exclusion of deprecated APIs, more discovery will be needed to converge on a specific technical strategy. It is currently thought this can be achieved by excluding classes/methods marked `deprecated` during JSII code generation. However, since the typescript/javascript packages are emitted by the typescript compiler, it will have to be done differently for the typscript language target. The current idea is to hide the deprecated APIs by removing their types from the `.d.ts` files emitted during build.

Reviewer: `/@deprecated/@internal/g` before compilation might do the trick

attention to which changes are scoped in for this release. The main criteria used to determine whether a change is
suitable for inclusion are:

- the change _requires_ breaking existing customers of the changed API or feature
  - a migration should be possible without incurring replacement of existing resources
- the change addresses a well documented customer pain point
- the change has had the due diligence done:
  - implications of the change are well documented
  - an implementation strategy is documented

## Changes considered for v2 inclusion

- The following changes will be included, or are candidates for inclusion:

  | Decision | RFC / Issue    | Description                                                                         |
  | -------- | -------------- | ----------------------------------------------------------------------------------- |
  | ✅ In    |                | Remove `@deprecated` APIs from the previous major version                           |
  | ✅ In    | [RFC-55]       | Reset all feature flags from the previous major version                             |

>   | ✅ In    | [RFC-6]        | Combine all AWS Construct Libraries into a single package                           |

Reviewer: ```suggestion
  | ✅ In    | [RFC-6]        | Monolithic packaging                           |
```

1. [Announce developer preview and document v1 maintenance plan](#developer-preview)
1. [Listen to customer feedback and adjust as needed](#feedback-phase)
1. [Announce General Availability](#general-availability)
1. [Clean up dead code](#clean-up)
1. [Monitor adoption of the new version](#monitoring-adoption)
1. [Deprecation of v1](#deprecation-of-v1)

## Forking v2 out of v1

The forking point will be achieved by cutting a new `next` branch from the current tip of the `master` branch. The
repository-wide `version` configuration will be updated to `2.0.0-alpha` (the pre-release identifier could be
different). A continuous integration pipeline will be configured on this new branch, so that pull requests can be
automatically validated.

As long as the codebases have not diverged too much, it should be possible to continue forward-porting new developments
on top of the `next` branch by simply cherry-picking new commits from the `master` branch. Alternatively, it might be
possible to simply periodically merge `master` into `next`.

As the codebases grow further appart, manual work may be required in order to forward-port feature work merged in

> `master`. As a consequence we will strive to keep divergences to a minimum for as long as possible.

Reviewer: This is going to be quite a hassle, so we need to set up some ownership around it. It won't happen on volunteer basis.

`@aws-cdk/stdlib` on the `master` branch. Availability of this library as part of CDK v1 is going to provide customers
with a stepping stone to begin migrating their applications to CDK v2.

Additionally, all modules that are re-packaged as part of `@aws-cdk/stdlib` will be marked as `"private": true` in the
`next` branch, making these available _only_ via `@aws-cdk/stdlib` in CDK v2.

## Documentation Updates

Before formally announcing the availability of CDK v2 for early adopters, a documentation source must be availble to
facilitate use of the new repository. The minimum bar should be availability of an API reference documentation for the
v2 codebase that is usable.

At the same time, the work to update the [AWS CDK User Guide] should be scoped and planned out. The updates to the User
Guide must also include a migration guide to help developers move from CDK v1 over to v2. Special attention will be
spent in ensuring the v2 release notes contain a clear and accurate summary of breaking changes and how users should
update their applications.

[aws cdk user guide]: https://docs.aws.amazon.com/cdk/latest/guide/home.html

> ## Prerelease Publishing

Reviewer: I think we should begin pre-release publishing as soon as we fork (or close to it) so it will be possible to test it in the wild. We don't have to announce it right away, but it's definitely something we should be able to point people to so they can give us early feedback. Trying to learn some lessons from our work on mono-cdk where we postponed this too long. For example, we will migrate delivlib and other internal projects to it.

There is no reason to wait for the documentation fix.

Once sufficient documentation is in place, and the framework-level changes have been implemented and integration tested,
we will be able to start promoting the upcoming new release and point early adopters to experimental builds of CDK v2.
This will allow gathering early feedback on the updated experience.

A prerelease announcement will be published, formally announcing our plan to release AWS CDK v2 in the near future. This
prerelease announcement should at least mention the following elements:

- A provisional list of breaking changes expected to be made to the _AWS Construct Library_ (`@aws-cdk/stdlib`) APIs
  during the [Construct Library Maintenance Tasks](#construct-library-maintenance-tasks) phase
- A finalized proposal for the CDK [v1 maintenance plan](#ongoing-maintenance-plan), and a call for developers to
  provide feedback on that
- A minimal migration guide for bravehearts who are inclined to try and migrate their existing (**non-production**)
  applications

> ## Construct Library Maintenance Tasks

Reviewer: I think I am leaning towards agreeing more and more with @rix0rrr on this. I think every breaking change needs to be vetted as part of this RFC and if it is approved it will be done. I am taking back my proposal to allow owners to just make breaking changes after the fork.

Reviewer 5: > I think every breaking change needs to be vetted as part of this RFC and if it is approved it will be done.

Given the aspirational timeline, I reckon this means no Construct Library breaking changes (beyond dropping `@deprecated` APIs) will be included in v2. I don't see how library owners will be able to afford the due diligence of an RFC before we need to have a delivery plan signed off.

Reviewer: We can give people a couple of weeks to propose minor breaking changes in their modules for review. Worst case we won't have any.

Reviewer: I am hopeful that most desired changes are already captured through deprecations so I am also fine with just picking those up.

Now that the v2 framework is in place, _AWS Construct Library_ owners (as well as third party contributors) will be
given some time to perform API-breaking maintenance tasks. The API-breaking changes should be limited to items that were

> scoped out during the [Prerelease Publishing](#prerelease-publishing) phase and listed in the prerelease announcepent,

Reviewer: ```suggestion
scoped out during the [Prerelease Publishing](#prerelease-publishing) phase and listed in the prerelease announcement,
```

however additional changes can be introduced if there is sufficient justification. All breaking changes (including those
planned in the prerelease announcement) will be documented in the release notes for each subsequent prerelease version,
per the standard process.

## Developer Preview

> At the end of the [Constrcut Library Maintenance Tasks](#construct-library-maintenance-tasks) window, a moratorium on

Reviewer: ```suggestion
At the end of the [Construct Library Maintenance Tasks](#construct-library-maintenance-tasks) window, a moratorium on
```

behavior_ ideally becomes the _only behavior_.

In certain cases, users may be unable to migrate over to the _new behavior_ of a feature flag without incurring
replacement of resources. Removing ability to toggle the feature off in a new major version would significantly
complicate upgrading applications to the new major version. Feature flags will be critically evaluated, and in cases
where a migration is impossible or difficult, the default state will be changed from disabled to enabled, but users will
retain the ability to disable as needed.

In order to reduce the codebase divergence between `master` and `next` (in order to minimize the amount of manual work
involved in forward-porting features on `next`), existing feature flags must be made version-aware, such that they are
enabled by default (and impossible to disable if a clean and reasonably simple migration is possible) when the current
version is greater then `2.0.0-0`.

The actual removal of code paths that support the _old behavior_ of feature flags that can no longer be disabled will be
deferred to _after_ the new major version has reached _General Availability_.

> ### Monolithic Packaging

Reviewer: Should be moved after the fork and before we start publishing

`monocdk-experiment` package has the tools necessary to re-package the existing _hypermodular_ libraries into a single
packaging. Re-using that same process will allow keeping the difference between `master` and `next` as constrained as
possible (reducing the need for manual intervention in forward-porting features).

Consequently, the `monocdk-experiment` package will be renamed so that it is published and advertised as
`@aws-cdk/stdlib` on the `master` branch. Availability of this library as part of CDK v1 is going to provide customers
with a stepping stone to begin migrating their applications to CDK v2.

Additionally, all modules that are re-packaged as part of `@aws-cdk/stdlib` will be marked as `"private": true` in the
`next` branch, making these available _only_ via `@aws-cdk/stdlib` in CDK v2.

## Documentation Updates

Before formally announcing the availability of CDK v2 for early adopters, a documentation source must be availble to
facilitate use of the new repository. The minimum bar should be availability of an API reference documentation for the
v2 codebase that is usable.

> 

Reviewer 5: Just the lines below:
- Finalize breaking changes to be expected
- Have signed off on the v1 maintenance plan
- Have a (preliminary) migration guide ready

> 

Reviewer: What I meant to ask is what does API reference documentation support for 2.0 looks like? It would be useful to have a rough idea of the scope of work required. Things that come to mind: add support for multiple versions in the web UI, add support for submodules.

The reason I think this would be useful is to de-risk the project - if we know that this is going to be a major piece of work, we need to start tackling it sooner rather than later since it's in the critical path of the project.

> 

Reviewer 6: I support the notion of the API reference for CDK v2.0 have a similar layout to the current version, i.e. a separate page for each service/module. 

We've heard so much feedback that the non-TypeScript portions of the API reference are sub-optimal, we really need to address the pain points in v2. Most of the feedback is for the API ref boils down to:

1. the overview page is not the default landing page when you navigate to a module in the Python API re
2. not all of the code snippets translated from TypeScript to the other languages are 100% correct (debugging incorrect sample code in official AWS docs is a poor customer experience), 
3. there are no code snippets or meaningful README content for "cfn-only" modules, and some customers don't know how to use the L1 constructs available in these modules - leading them to think the module/service is not yet available in the CDK

  - [ ] API Reference Documentation
  - [ ] User Guide
  - [ ] Migration Guide
  - [ ] Final maintenance plan for v1
- [ ] Quality Assurance
  - [ ] Complete suite of integration tests
  - [ ] Examples built and tested against v2
- [ ] No v1 feature is missing from v2 (unless it was deprecated in v1)

Once the criteria have been satisfied, the high-level procedure for officially releasing CDK v2 is:

1. Create a new `v1` branch from the `master` branch
   1. Configure GitHub branch protection on `v1` as appropriate
   1. Deploy CI/CD infrastructure on top of `v1`
1. Make `next` the new `master` branch:
   1. Force-push `next` on `master`
   1. Delete `next` and all related CI/CD infrastructure
1. Perform a _normal_ release
1. Ceremoniously annouce CDK v2 is _Generally Available_, and repeat the v1 maintenance plan

## Clean Up

Now that the new major version has been declared _Generally Available_, the previous major (`v1`) has entered
_maintenance_ mode. Forward porting is no longer exercised: instead, bug fixes on `master` will be back-ported to `v1`
when necessary. This means `master` can diverge more significantly from `v1` than in earler stages of the process.

> The definition of `v1` feature flags and the code paths supporthing the _old behavior_ of those can be deleted.

Reviewer 5: No the feature flags can be *neutralized* in `v2` (i.e: you CANNOT disable them), but the code paths must be kept in order to avoid tricky merge conflicts.

> The definition of `v1` feature flags and the code paths supporthing the _old behavior_ of those can be deleted.

Reviewer: Okay got it. Makes sense. It wasn't super clear to me what "neutralized" means.

## Monitoring Adoption

We will keep an eye on adoption of CDK v2 thanks to the [version reporting] dataset. Should adoption pick up too slowly,
actions will be taken to understand the causes for slow adoption and identify opportunities to provide additional
tooling to remove roadblocks and facilitate migration.

[version reporting]: https://docs.aws.amazon.com/cdk/latest/guide/tools.html

## Deprecation of v1

Based on the v1 maintenance plan, feature development is to stop on CDK v1, as only bugs reported against CDK v1 are to

> be fixed. Back-porting of features is not necessary, although the community is still welcome to back-port features.

Reviewer: Is this the plan? To fix all bugs reported against 1.x both on 1.x and on 2.x? That's going to be pretty expensive. This is something we need to be very intentional about.

Reviewer 5: Either the bug exists in v1 AND v2, and then the fix is likely the same on both sides. Ergo I am assuming it will not be "significantly" more expensive to fix both.

Or the bug exists only in v1 and there is nothing to fix in v2.

  changes. Once AWS identifies a release to be a stable product, it may mark it as a _Release Candidate_. _Release
  Candidates_ are ready for _General Availability_ unless significant bugs emerge, and will receive full AWS support.
- **Full Support**: During this phase, software is fully supported. AWS will provide regular releases that include new
  features, support for new services, API updates for existing services, as well as bug and security fixes.
- **Maintenance Announcement**: AWS will make a public announcement at least 6 months before software enters the
  **Maintenance** phase. During this period, software will continue to be fully supported. Typically, this announcement
  happens at the same time as the next major version is introduced.
- **Maintenance**: During this phase, AWS limits releases to address critical bug fixes and security issues only.
  Software will no longer receive API updates for new or existing services, or be updated to support new regions. The
  **Maintenance** phase has a default duration of 12 months, unless otherwise specified as part of the **Maintenance

>   Announcement**.

Reviewer: okay, this does *not* imply that any bug reported on 1.x will also be fixed on 1.x

Reviewer 5: Only if it's critical / security impacting, yes.

- **End-of-Support**: This phase starts at the scheduled end of the **Maintenance** phase. Software that has reached
  **End-of-Support** will no longer receive updates or new releases. Previously published releases will continue to be
  available via public package managers and the code will remain accessible on GitHub. The GitHub repository may be
  archived. Use of software that has reached **End-of-Support** is done at the user's discretion. We recommend users
  upgrade to the latest major version (the release in **Full Support** phase).

Here is a visual illystration of the major version life-cycle (timelines are purely illustrative):

![Illustration of the Maintenance Policy](../images/MaintenancePolicy.png)

# Rationale and Alternatives

## Rationale

The main motivation for moving forward with a v2 release now is the desire to start moving towards the single module
structure. The current multi module structure is known to cause a lot of confusion for users. For more details on why
aws-cdk is moving to a single package, see the [MonoCDK RFC](https://github.com/aws/aws-cdk-rfcs/issues/6).

> This belief that releasing a single package will make user's lives easier is countered by a strong desire to limit the

Reviewer: The main motivation for mono-cdk is to allow 3rd party libraries to declare CDK as a peer dependency without causing breaking changes any time additional transitive dependencies are added upstream. Basically monolithic packaging is the only viable solution.

Reviewer 5: This paragraph isn't about "why monoCDK" (this is explained in the MonoCDK RFC as pointed to the line before), but more "why is this v2" / "why is v2 the way it is"

problems currently experienced by users caused by the current packaging strategy and these problems will always exist
unless we stop supporting the `@aws-cdk/` packages.

Since we want to eventually supplant the current module structure with the new one, semver says we have to do it with
the release of a major version.

# Adoption Strategy

All users will perform the following to adopt CDK v2

1. Update Dependencies - Changing from referencing the old `@aws-cdk/` packages to the new `@aws-cdk/stdlib` package in
   the corresponding target language's package manifest (`package.json`, `.csproj`, `pom.xml`, `setup.py`, etc...).
1. Update Imports - Since the current module structure is being changed in certain languages, import statements in
   user's code will need to change accordingly:
   - In TypeScript, this can be automated by a tool such as `@monocdk-experiment/import-rewriter`
   - In Python, manual replacement of import statements migth be necessary
   - In .NET languages (C#, F#, ...) and Java, the namespace structure from CDK v1 will be retained, removing the need
     for any code change

> 

Reviewer: link to the construct compat layer removal RFC which has more details on breaking changes

  - `@aws-cdk/aws-ram`
  - `@aws-cdk/aws-rds`
  - `@aws-cdk/aws-redshift`
  - `@aws-cdk/aws-resourcegroups`
  - `@aws-cdk/aws-robomaker`
  - `@aws-cdk/aws-route53-patterns`
  - `@aws-cdk/aws-route53-targets`
  - `@aws-cdk/aws-route53`
  - `@aws-cdk/aws-route53resolver`
  - `@aws-cdk/aws-s3-assets`
  - `@aws-cdk/aws-s3-deployment`
  - `@aws-cdk/aws-s3-notifications`
  - `@aws-cdk/aws-s3`
  - `@aws-cdk/aws-sagemaker`
  - `@aws-cdk/aws-sam`
  - `@aws-cdk/aws-sdb`
  - `@aws-cdk/aws-secretsmanager`
  - `@aws-cdk/aws-securityhub`
  - `@aws-cdk/aws-servicecatalog`
  - `@aws-cdk/aws-servicediscovery`
  - `@aws-cdk/aws-ses-actions`
  - `@aws-cdk/aws-ses`
  - `@aws-cdk/aws-sns-subscriptions`
  - `@aws-cdk/aws-sns`
  - `@aws-cdk/aws-sqs`
  - `@aws-cdk/aws-ssm`
  - `@aws-cdk/aws-stepfunctions-tasks`
  - `@aws-cdk/aws-stepfunctions`
  - `@aws-cdk/aws-synthetics`
  - `@aws-cdk/aws-transfer`
  - `@aws-cdk/aws-waf`
  - `@aws-cdk/aws-wafregional`
  - `@aws-cdk/aws-wafv2`
  - `@aws-cdk/aws-workspaces`
- `@aws-cdk/app-delivery`
- `@aws-cdk/assets`
- `@aws-cdk/cloud-assembly-schema`
- `@aws-cdk/cloudformation-include`
- `@aws-cdk/custom-resource`
- `@aws-cdk/cx-api`
- `@aws-cdk/pipelines`

> - `@aws-cdk/region-info`

Reviewer: All of these (minor `app-delivery` which is deprecated) should be included.

---
feature name: cdk-2.0
start date: 2020-3-29
rfc pr: https://github.com/aws/aws-cdk-rfcs/pull/156
related issue: https://github.com/aws/aws-cdk-rfcs/issues/6
---

<!-- replace the blockquoted sections with your content -->

# Summary

This RFC details the strategy for building and releasing v2 of the AWS CDK, in addition to changes to tooling required
to support this strategy. For details on the specific features themselves, refer to the corresponding RFCs.

# Motivation

> Since the CDK was announcned as "generally available", the team has tried to limit breaking changes for users. However,

Reviewer 7: ```suggestion
Since the CDK was announced as "generally available", the team has tried to limit breaking changes for users. However,
```

Guide must also include a migration guide to help developers move from CDK v1 over to v2. Special attention will be
spent in ensuring the v2 release notes contain a clear and accurate summary of breaking changes and how users should
update their applications.

[aws cdk reference documentation]: https://docs.aws.amazon.com/cdk/api/latest/
[aws cdk user guide]: https://docs.aws.amazon.com/cdk/latest/guide/home.html

## Prerelease Announcement

Once sufficient documentation is in place, and the framework-level changes have been implemented and integration tested,
we will be able to start promoting the upcoming new release and point early adopters to experimental builds of CDK v2.
This will allow gathering early feedback on the updated experience.

A prerelease announcement will be published, formally announcing our plan to release AWS CDK v2 in the near future. This
prerelease announcement should at least mention the following elements:

- A provisional list of breaking changes expected to be made to the _AWS Construct Library_ (`aws-cdk-lib`) APIs during
  the [Construct Library Maintenance Tasks](#construct-library-maintenance-tasks) phase

> - A finalized proposal for the CDK [v1 maintenance plan](#ongoing-maintenance-plan), and a call for developers to

Reviewer 7: ```suggestion
- A finalized proposal for the CDK [v1 maintenance plan](#aws-cdk-maintenance-policy), and a call for developers to
```

## Monitoring Adoption

We will keep an eye on adoption of CDK v2 thanks to the [version reporting] dataset. Should adoption pick up too slowly,
actions will be taken to understand the causes for slow adoption and identify opportunities to provide additional
tooling to remove roadblocks and facilitate migration.

[version reporting]: https://docs.aws.amazon.com/cdk/latest/guide/tools.html

## Deprecation of v1

Based on the [maintenance policy](#aws-cdk-maintenance-policy), new feature development is to stop on CDK v1, and only
critical bugs and security issues will be addressed in the `v1` release line.

New GitHub labels will be added to differentiate bugs reported against `v1` from those reported against `v2`. Bugs
reported against `v1` will be closed with a standard message during triaging unless they are identified as critical or
security-impacting (meaning they qualify for `P0` handling).

A similar provision applies to pull-requests, as they will only be accepted in the `v1` branch if they are non-breaking
and fall into one of the following categories:

- features dedicated to facilitating migration from `v1` to `v2`
- features back-ported from `v2`

> - bug fixes (where relevant, the author is expected to submit the `v2` fix at the same time)

Reviewer 7: ```suggestion
- `P0` bug fixes (where relevant, the author is expected to submit the `v2` fix at the same time)
```

Is this correct (based on the above paragraph)? We will only accept `P0` bugs on `v1`, so we'll only accept PRs for `P0` bug fixes?

Reviewer 5: No - I think we should accept bug fixes from the community for any severity, but we (the core team) will not author them.

of time and warning to change their code and provide details about a reasonable path to do so without interruptions to
their existing cloud resources.

---

# Annexes

## Annex 1

Packages that are included in the monolithic library bundle (`monocdk-experiment` to become `aws-cdk-lib`) are the
following:

- All non-deprecated packages whith a name starting with `@aws-cdk/aws-` or `@aws-cdk/alexa-`:
  - `@aws-cdk/alexa-ask`
  - `@aws-cdk/aws-accessanalyzer`
  - `@aws-cdk/aws-acmpca`
  - `@aws-cdk/aws-amazonmq`
  - `@aws-cdk/aws-amplify`
  - `@aws-cdk/aws-apigateway`
  - `@aws-cdk/aws-apigatewayv2`
  - `@aws-cdk/aws-appconfig`
  - `@aws-cdk/aws-applicationautoscaling`
  - `@aws-cdk/aws-appmesh`
  - `@aws-cdk/aws-appstream`
  - `@aws-cdk/aws-appsync`
  - `@aws-cdk/aws-athena`
  - `@aws-cdk/aws-autoscaling-common`
  - `@aws-cdk/aws-autoscaling-hooktargets`
  - `@aws-cdk/aws-autoscaling`
  - `@aws-cdk/aws-autoscalingplans`
  - `@aws-cdk/aws-backup`
  - `@aws-cdk/aws-batch`
  - `@aws-cdk/aws-budgets`
  - `@aws-cdk/aws-cassandra`
  - `@aws-cdk/aws-ce`
  - `@aws-cdk/aws-certificatemanager`
  - `@aws-cdk/aws-chatbot`
  - `@aws-cdk/aws-cloud9`
  - `@aws-cdk/aws-cloudformation`

>   - `@aws-cdk/aws-cloudfront`
>   - `@aws-cdk/aws-cloudtrail`

Reviewer 7: ```suggestion
  - `@aws-cdk/aws-cloudfront`
  - `@aws-cdk/aws-cloudfront-origins`
  - `@aws-cdk/aws-cloudtrail`
```

  | ✅ In    |             | Remove `@deprecated` APIs from the previous major version                           |
  | ✅ In    | [RFC-55]    | Reset all feature flags from the previous major version                             |
  | ✅ In    | [RFC-6]     | Combine all AWS Construct Libraries into a single package                           |
  | ✅ In    | [RFC-192]   | Remove the _constructs compatibility layer_ in favor of using `constructs` directly |

* The following changes were scoped out because a solution is likely to be achievable without breaking existing
  customers:

  | Decision | RFC / Issue | Description                                                                |
  | -------- | ----------- | -------------------------------------------------------------------------- |
  | 🚫 Out   | [#39]       | Distribute lambda code for bundled custom resources via a public S3 bucket |
  | 🚫 Out   | [RFC-77]    | Generate CloudFormation Resources from the CloudFormation Registry schemas |
  | 🚫 Out   | [RFC-193]   | Fix API for TypeScript type unions when used in statically typed languages |

>   | 🚫 Out   | [#116]      | Easier identification of `@experimental` modules and APIs                  |

Reviewer 6: What is the plan for "experimental" modules and APIs in CDK v2.0? Too many customers have come to the wrong conclusion about the stability of the CDK because they were using experimental modules and didn't know it; presumably because they didn't read the API reference labels. The note above this table states the solution is "likely" achievable without breaking customers, but how we solve this particular problem is not stated. I like the proposal in issue #116 to have 2 versions of the CDK; one stable and one stable+experimental.

Reviewer 2: The non-breaking thing we can do here quickly is add `"(experimental)"` to the front of every experimental API's docstring.

Two different libs won't work well as soon as we involve 3rd party libraries.

- [ ] No v1 feature is missing from v2 (unless it was deprecated in v1)

Once the criteria have been satisfied, the high-level procedure for officially releasing CDK v2 is:

1. Create a new `v1` branch from the `master` branch
   1. Configure GitHub branch protection on `v1` as appropriate
   1. Deploy CI/CD infrastructure on top of `v1`
1. Make `next` the new `master` branch:
   1. Force-push `next` on `master`
   1. Delete `next` and all related CI/CD infrastructure
1. Perform a _normal_ release
1. Ceremoniously annouce CDK v2 is _Generally Available_, and repeat the v1 maintenance plan

## Clean Up

Now that the new major version has been declared _Generally Available_, the previous major (`v1`) has entered
_maintenance_ mode. Forward porting is no longer exercised: instead, bug fixes on `master` will be back-ported to `v1`
when necessary. This means `master` can diverge more significantly from `v1` than in earler stages of the process.

The definition of `v1` feature flags and the code paths supporthing the _old behavior_ of those can be deleted (they
were previously simply _neutralized_ in `v2` so that they could no longer be disabled there).

## Monitoring Adoption

> 

Reviewer 6: I don't think there is much work to do on the CDK CLI side of our metrics collection, but we'll need to rethink how we use the AWS::CDK::Metadata resource to track usage since all CDK constructs will be delivered in a single package/module. Ideally, we'll get more granular data on the actual constructs (L1, L2, or L3) used in CDK code.

[MonoCDK RFC](https://github.com/aws/aws-cdk-rfcs/issues/6).

The second is to simply release the new module, `aws-cdk-lib` locked to the same `1.x.x` version that the other modules
are currently locked to. In fact, that is the first stage of the release process, its just under the name
`monocdk-experiment`. This was a low cost way for the dev team, contributors, and users to experiment with the new
structure before investing too much towards it.

Regardless, the point of using this change to push towards a v2 release, is that the single package structure is "the
new cdk". Meaning, the long term plan is to only release the `aws-cdk-lib` package and no longer publish the other
`@aws-cdk/` namespaced ones. The [MonoCDK RFC](https://github.com/aws/aws-cdk-rfcs/issues/6) has more info on the
problems currently experienced by users caused by the current packaging strategy and these problems will always exist
unless we stop supporting the `@aws-cdk/` packages.

Since we want to eventually supplant the current module structure with the new one, semver says we have to do it with
the release of a major version.

# Adoption Strategy

All users will perform the following to adopt CDK v2

> 

Reviewer 6: What's the feasibility of a migration tool to help customers update their dependencies and imports?

Reviewer 5: All of this is *easy* if you consider **one** language (**TypeScript**/**Javascript**). Things become **much** hairier once you start to think about **Python**, **C#**, **Java**, ... I think if we go into tooling we should put ourselves in the position to offer an equally managed experience for **all** languages (and I don't think this is realistic).
